File: kirun-js/src/engine/function/system/array/LastIndexOf.ts
import { KIRuntimeException } from '../../../exception/KIRuntimeException';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { PrimitiveUtil } from '../../../util/primitive/PrimitiveUtil';
import { AbstractArrayFunction } from './AbstractArrayFunction';

export class LastIndexOf extends AbstractArrayFunction {
    public constructor() {
        super(
            'LastIndexOf',
            [
                LastIndexOf.PARAMETER_ARRAY_SOURCE,
                LastIndexOf.PARAMETER_ANY_ELEMENT_OBJECT,
                LastIndexOf.PARAMETER_INT_FIND_FROM,
            ],
            LastIndexOf.EVENT_RESULT_INTEGER,
        );
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let source: any[] = context
            ?.getArguments()
            ?.get(LastIndexOf.PARAMETER_ARRAY_SOURCE.getParameterName());

        let find = context
            ?.getArguments()
            ?.get(LastIndexOf.PARAMETER_ANY_ELEMENT_OBJECT.getParameterName());

        let len = context
            ?.getArguments()
            ?.get(LastIndexOf.PARAMETER_INT_FIND_FROM.getParameterName());

        if (source.length == 0)
            return new FunctionOutput([
                EventResult.outputOf(new Map([[LastIndexOf.EVENT_RESULT_NAME, -1]])),
            ]);

        if (len < 0 || len > source.length)
            throw new KIRuntimeException(
                "The value of length shouldn't the exceed the size of the array or shouldn't be in terms",
            );

        let index: number = -1;

        for (let i: number = source.length - 1; i >= len; i--) {
            if (PrimitiveUtil.compare(source[i], find) == 0) {
                index = i;
                break;
            }
        }

        return new FunctionOutput([
            EventResult.outputOf(new Map([[LastIndexOf.EVENT_RESULT_NAME, index]])),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/array/last_index_of.py
from __future__ import annotations

from typing import Any, List, TYPE_CHECKING

from kirun_py.exception.ki_runtime_exception import KIRuntimeException
from kirun_py.function.system.array.abstract_array_function import AbstractArrayFunction
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.util.primitive.primitive_util import PrimitiveUtil

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class LastIndexOf(AbstractArrayFunction):

    def __init__(self) -> None:
        super().__init__(
            'LastIndexOf',
            [
                AbstractArrayFunction.PARAMETER_ARRAY_SOURCE,
                AbstractArrayFunction.PARAMETER_ANY_ELEMENT_OBJECT,
                AbstractArrayFunction.PARAMETER_INT_FIND_FROM,
            ],
            AbstractArrayFunction.EVENT_RESULT_INTEGER,
        )

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        source: List[Any] = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.get_parameter_name()
        )
        find = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_ANY_ELEMENT_OBJECT.get_parameter_name()
        )
        from_idx: int = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_INT_FIND_FROM.get_parameter_name()
        )

        if len(source) == 0:
            return FunctionOutput([
                EventResult.output_of(
                    {AbstractArrayFunction.EVENT_RESULT_NAME: -1}
                )
            ])

        if from_idx < 0 or from_idx > len(source):
            raise KIRuntimeException(
                "The value of length shouldn't the exceed the size of the array "
                "or shouldn't be in terms"
            )

        index: int = -1
        for i in range(len(source) - 1, from_idx - 1, -1):
            if PrimitiveUtil.compare(source[i], find) == 0:
                index = i
                break

        return FunctionOutput([
            EventResult.output_of(
                {AbstractArrayFunction.EVENT_RESULT_NAME: index}
            )
        ])
